<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Tools to Work with HLA Typings and Single-Antigen Bead Assays • hlapro</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Tools to Work with HLA Typings and Single-Antigen Bead Assays">
<meta name="description" content="Provides tools for Human Leukocyte Antigen (HLA) data: genotypes as well as (Luminex) single-antigen bead-based assays. Including: standardizing HLA typings to a common format (GL Strings), reducing and increasing HLA typing resolution, determining bead positivity and assessing donor-specific antibodies (DSAs).">
<meta property="og:description" content="Provides tools for Human Leukocyte Antigen (HLA) data: genotypes as well as (Luminex) single-antigen bead-based assays. Including: standardizing HLA typings to a common format (GL Strings), reducing and increasing HLA typing resolution, determining bead positivity and assessing donor-specific antibodies (DSAs).">
<meta property="og:image" content="https://lcreteig.github.io/hlapro/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">hlapro</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/lcreteig/hlapro/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="hlapro-">hlapro 
<a class="anchor" aria-label="anchor" href="#hlapro-"></a>
</h1>
</div>
<!-- badges: start -->

<p>The goal of hlapro is to provide some tooling to work with <a href="https://en.wikipedia.org/wiki/Human_leukocyte_antigen" class="external-link">Human Leukocyte Antigen (HLA)</a> data, in the context of transplantation research:</p>
<ul>
<li>HLA typings</li>
<li>single bead assays (e.g. Luminex) for detecting HLA-antibodies</li>
</ul>
<p>The package is currently still in (heavy) development, but ultimately the goal is to support the following functionality:</p>
<ul>
<li>
<strong>HLA typings</strong>:
<ul>
<li>Obtaining HLA mismatches between transplant donor and recipient</li>
<li>Formatting into <a href="https://glstring.org" class="external-link">GL Strings</a>
</li>
<li>Downscaling (reducing) typing resolution</li>
<li>Upscaling (imputing) typing resolution, based on haplotype frequency tables</li>
</ul>
</li>
<li>
<strong>HLA antibody assays</strong>:
<ul>
<li>Determining which beads in the assay are positive, based on automated implementations of different methods</li>
<li>Assessing the presence of donor-specific antibodies (DSA)</li>
</ul>
</li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of hlapro from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"lcreteig/hlapro"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lcreteig/hlapro" class="external-link">hlapro</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="typings">Typings<a class="anchor" aria-label="anchor" href="#typings"></a>
</h3>
<div class="section level4">
<h4 id="mismatches">Mismatches<a class="anchor" aria-label="anchor" href="#mismatches"></a>
</h4>
<p>Get mismatched HLAs from donor and recipient typings</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">donor_typing</span> <span class="op">&lt;-</span> <span class="st">"A1 A2 B5"</span></span>
<span><span class="va">recipient_typing</span> <span class="op">&lt;-</span> <span class="st">"A1 A3 B5 B12"</span></span>
<span></span>
<span><span class="fu"><a href="reference/get_mismatches.html">get_mismatches</a></span><span class="op">(</span><span class="va">donor_typing</span>, <span class="va">recipient_typing</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A2"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extracting-alleles">Extracting alleles<a class="anchor" aria-label="anchor" href="#extracting-alleles"></a>
</h4>
<p>Extract HLA alleles from a typing string</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing</span> <span class="op">&lt;-</span> <span class="st">"A1 A2 B7 B8 Cw3 DQ5 DQ8 DR4 DR11 DR52 DR53"</span></span>
<span><span class="fu"><a href="reference/extract_alleles_str.html">extract_alleles_str</a></span><span class="op">(</span><span class="va">typing</span><span class="op">)</span></span>
<span><span class="co">#&gt;    A_1    A_2    B_1    B_2    C_1    C_2 DPA1_1 DPA1_2 DPB1_1 DPB1_2 DQA1_1 </span></span>
<span><span class="co">#&gt;    "1"    "2"    "7"    "8"    "3"     NA     NA     NA     NA     NA     NA </span></span>
<span><span class="co">#&gt; DQA1_2 DQB1_1 DQB1_2 DRB1_1 DRB1_2 DRB._1 DRB._2 </span></span>
<span><span class="co">#&gt;     NA    "5"    "8"    "4"   "11"   "52"   "53"</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>typing <span class="op">=</span> <span class="st">"A1 A2 B7 B8 Cw3 DQ5 DQ8 DR4 DR11 DR52 DR53"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/extract_alleles_str.html">extract_alleles_df</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">typing</span>, loci <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(typing)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(typing)`</span></span>
<span><span class="co">#&gt; # A tibble: 1 × 7</span></span>
<span><span class="co">#&gt;   typing                                     A_1   A_2   B_1   B_2   C_1   C_2  </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                                      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;</span></span>
<span><span class="co">#&gt; 1 A1 A2 B7 B8 Cw3 DQ5 DQ8 DR4 DR11 DR52 DR53 1     2     7     8     3     ""</span></span></code></pre></div>
<p>If there’s more than two alleles for a given locus in the typing, you’ll receive a warning:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing</span> <span class="op">&lt;-</span> <span class="st">"A1 A2 A3 B7 B8 Cw1"</span></span>
<span><span class="fu"><a href="reference/extract_alleles_str.html">extract_alleles_str</a></span><span class="op">(</span><span class="va">typing</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: One or more loci found with more than 2 alleles.</span></span>
<span><span class="co">#&gt; ✖ `extract_alleles_str()` will only pick the first two.</span></span>
<span><span class="co">#&gt; ℹ Use `hlapro::count_alleles()` to find out more.</span></span>
<span><span class="co">#&gt;    A_1    A_2    B_1    B_2    C_1    C_2 DPA1_1 DPA1_2 DPB1_1 DPB1_2 DQA1_1 </span></span>
<span><span class="co">#&gt;    "1"    "2"    "7"    "8"    "1"     NA     NA     NA     NA     NA     NA </span></span>
<span><span class="co">#&gt; DQA1_2 DQB1_1 DQB1_2 DRB1_1 DRB1_2 DRB._1 DRB._2 </span></span>
<span><span class="co">#&gt;     NA     NA     NA     NA     NA     NA     NA</span></span></code></pre></div>
<p>Use <code><a href="reference/count_alleles.html">count_alleles()</a></code> to easily inspect the number of alleles per locus:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/count_alleles.html">count_alleles</a></span><span class="op">(</span><span class="va">typing</span><span class="op">)</span></span>
<span><span class="co">#&gt;    A    B    C DPA1 DPB1 DQA1 DQB1 DRB1 DRB. </span></span>
<span><span class="co">#&gt;    3    2    1    0    0    0    0    0    0</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cleaning-allele-strings">Cleaning allele strings<a class="anchor" aria-label="anchor" href="#cleaning-allele-strings"></a>
</h4>
<p>Correcting some common formatting issues</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allele_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">" A'*1 "</span>, <span class="co"># spurious whitespace, punctuation, no leading zero, missing XX code</span></span>
<span>  <span class="st">"A10(25)"</span>, <span class="co"># redundant typing (includes both split and broad)</span></span>
<span>  <span class="st">"C*03:01/02"</span> <span class="co"># locus and allele group left off in ambiguity</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/clean_hla.html">clean_hla</a></span><span class="op">(</span><span class="va">allele_vec</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A*01:XX"         "A25"             "C*03:01/C*03:02"</span></span></code></pre></div>
<p><code><a href="reference/clean_hla.html">clean_hla()</a></code> can also convert typings in the pre-2010 (v2) nomenclature to the updated (v3) nomenclature:…</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/clean_hla.html">clean_hla</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cw*030205"</span>, <span class="st">"A*2416"</span>, <span class="st">"DPB1*02BYVD"</span>, <span class="st">"B*35UMU"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "C*03:02:05"   "A*31:08"      "DPB1*02:FNWG" "B*35:WRE"</span></span></code></pre></div>
<p>…and convert deleted/changed alleles to their new assignments:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/clean_hla.html">clean_hla</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C*03:12"</span>, <span class="st">"C*12:139"</span>, <span class="st">"A*02:100"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "C*03:19"   "C*12:139Q" NA</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="validating-alleles">Validating alleles<a class="anchor" aria-label="anchor" href="#validating-alleles"></a>
</h4>
<p>Check whether alleles are well-formed</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/validate_allele.html">validate_allele</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A2"</span>, <span class="st">"A*01:AABJE"</span>, <span class="st">"A*24:02:01:02L"</span>, <span class="st">"not-an-HLA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  TRUE  TRUE  TRUE FALSE</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="allele-resolution">Allele resolution<a class="anchor" aria-label="anchor" href="#allele-resolution"></a>
</h4>
<p>Determine whether an allele is of low/intermediate/high resolution</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/get_resolution.html">get_resolution</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A2"</span>, <span class="st">"A*01:AABJE"</span>, <span class="st">"B*42:08"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "low"          "intermediate" "high"</span></span></code></pre></div>
<p>Use the extended mode to get more information:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/get_resolution.html">get_resolution</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A2"</span>, <span class="st">"A*24:XX"</span>, <span class="st">"A*01:AB"</span>, <span class="st">"B*42:08"</span>, <span class="st">"A*01:01:01:01"</span><span class="op">)</span>,</span>
<span>  extended <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "serology - broad"    "molecular - split"   "intermediate"       </span></span>
<span><span class="co">#&gt; [4] "high - second field" "high - fourth field"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="downscaling-to-serological-equivalents">Downscaling to serological equivalents<a class="anchor" aria-label="anchor" href="#downscaling-to-serological-equivalents"></a>
</h4>
<p>There’s two ways to do this:</p>
<ol style="list-style-type: decimal">
<li>Using the <a href="https://www.ebi.ac.uk/ipd/imgt/hla/" class="external-link">IPD-IMGT/HLA database</a> as implemented by the <a href="https://network.nmdp.org/services-support/bioinformatics-immunobiology/tools" class="external-link">NMDP</a> in <a href="https://github.com/nmdp-bioinformatics/py-ard" class="external-link">py-ard</a>:</li>
</ol>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ard</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/db_initialize.html">db_initialize</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/path.expand.html" class="external-link">path.expand</a></span><span class="op">(</span><span class="st">"~/ipd_db"</span><span class="op">)</span><span class="op">)</span> <span class="co"># (download &amp;) initialize database</span></span>
<span><span class="fu"><a href="reference/reduce_to_serology.html">reduce_to_serology</a></span><span class="op">(</span></span>
<span>  <span class="va">ard</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B*15:79N"</span>, <span class="st">"B*15:25/B*15:61"</span>, <span class="st">"B*15:02"</span>, <span class="st">"B*15:01:16"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] ""            "B15/B62/B70" "B15/B70/B75" "B15/B62"</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Get the serological equivalents of an allele as defined by the <a href="https://etrl.eurotransplant.org/resources/hla-tables/" class="external-link">ETRL HLA</a> conversion tables. This is less extensive and precise, but is nonetheless relevant for the Eurotransplant region.</li>
</ol>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/get_serology.html">get_serology</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B*15:79N"</span>, <span class="st">"B*15:25/B*15:61"</span>, <span class="st">"B*15:02"</span>, <span class="st">"B*15:01:16"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] NA    "B15" "B75" "B62"</span></span></code></pre></div>
<p>The latter also supports lookup of broads or splits specifically:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alleles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*01:01"</span>, <span class="st">"A*25:76:02"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/get_split.html">get_split</a></span><span class="op">(</span><span class="va">alleles</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] NA    "A25"</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A24"</span>, <span class="st">"A*25:76:02"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/get_broad.html">get_broad</a></span><span class="op">(</span><span class="va">splits</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A9"  "A10"</span></span></code></pre></div>
<p>And whether an allele has the Bw4 or Bw6 epitope:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B14"</span>, <span class="st">"B63"</span>, <span class="st">"B*40:05"</span>, <span class="st">"A*01:01"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/get_public.html">get_public</a></span><span class="op">(</span><span class="va">b_s</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Bw6" "Bw4" "Bw6" NA</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="reducing-to-two-field-resolution">“Reducing” to two-field resolution<a class="anchor" aria-label="anchor" href="#reducing-to-two-field-resolution"></a>
</h4>
<p>For many applications it suffices to work with HLA alleles at the protein-level. <code>py-ard</code> can also leverage the IPD-IMGT/HLA database to accomplish this:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/reduce_to_field2.html">reduce_to_field2</a></span><span class="op">(</span><span class="va">ard</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*01:01:01:01"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A*01:01"</span></span></code></pre></div>
<p>The above case is arguably trivial, but many others are not:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/reduce_to_field2.html">reduce_to_field2</a></span><span class="op">(</span><span class="va">ard</span>, <span class="st">"A*01:04:01:01N"</span><span class="op">)</span> <span class="co"># shortnull</span></span>
<span><span class="co">#&gt; [1] "A*01:04N"</span></span>
<span><span class="fu"><a href="reference/reduce_to_field2.html">reduce_to_field2</a></span><span class="op">(</span><span class="va">ard</span>, <span class="st">"B*15:AH"</span><span class="op">)</span> <span class="co"># MACs</span></span>
<span><span class="co">#&gt; [1] "B*15:01/B*15:07"</span></span>
<span><span class="fu"><a href="reference/reduce_to_field2.html">reduce_to_field2</a></span><span class="op">(</span><span class="va">ard</span>, <span class="st">"Cw10"</span><span class="op">)</span> <span class="co"># serology</span></span>
<span><span class="co">#&gt; [1] "C*03:02/C*03:04Q/C*03:04/C*03:06/C*03:26/C*03:28/C*03:46"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="upscaling-from-serological-equivalents-to-2-field-high-resolution">Upscaling from serological equivalents to 2-field high resolution<a class="anchor" aria-label="anchor" href="#upscaling-from-serological-equivalents-to-2-field-high-resolution"></a>
</h4>
<p>Depends on the haplotype frequencies released by the NMDP, which must be downloaded from <a href="https://frequency.nmdp.org" class="external-link">here</a> after logging in and accepting the license.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/upscale_typings.html">upscale_typings</a></span><span class="op">(</span></span>
<span>  filepath <span class="op">=</span> <span class="st">"~/Downloads/A~C~B~DRB3-4-5~DRB1~DQB1.xlsx"</span>,</span>
<span>  typing <span class="op">=</span> <span class="st">"A24 A28 B35 B61 DR4 DR11"</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">unphased_geno</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"haplo"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 5</span></span>
<span><span class="co">#&gt;   unphased_geno              haplo_freq_1 haplo_freq_2 haplo_rank_1 haplo_rank_2</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                             &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 A*24:02g A*68:01g B*35:03…    0.000585     0.000209           240          665</span></span>
<span><span class="co">#&gt; 2 A*24:02g A*68:01g B*35:03…    0.0000849    0.0000597         1565         2058</span></span></code></pre></div>
<p>Also able to upscale multiple typings at once, for instance in a dataframe:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing_df</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"001"</span>, <span class="st">"002"</span><span class="op">)</span>,</span>
<span>  input_typings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"A24 A28 B35 B61 DR4 DR11"</span>,</span>
<span>    <span class="st">"A2 A3 B52 B35 Cw4 DR11 DR52 DQ3"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">typing_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>geno_df <span class="op">=</span> <span class="fu"><a href="reference/upscale_typings.html">upscale_typings</a></span><span class="op">(</span></span>
<span>    <span class="st">"~/Downloads/A~C~B~DRB3-4-5~DRB1~DQB1.xlsx"</span>,</span>
<span>    <span class="va">input_typings</span>,</span>
<span>    as_list <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/unnest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">geno_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 13</span></span>
<span><span class="co">#&gt;   id    input_typings id_unphased_geno unphased_geno unphased_freq unphased_prob</span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;                    &lt;int&gt; &lt;chr&gt;                 &lt;dbl&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 001   A24 A28 B35 …              364 A*24:02g A*6…  0.000000254          0.121</span></span>
<span><span class="co">#&gt; 2 001   A24 A28 B35 …              364 A*24:02g A*6…  0.000000254          0.121</span></span>
<span><span class="co">#&gt; 3 002   A2 A3 B52 B3…               17 A*02:01g A*0…  0.0000000542         0.151</span></span>
<span><span class="co">#&gt; 4 002   A2 A3 B52 B3…               17 A*02:01g A*0…  0.0000000542         0.151</span></span>
<span><span class="co">#&gt; 5 002   A2 A3 B52 B3…               17 A*02:01g A*0…  0.0000000542         0.151</span></span>
<span><span class="co">#&gt; 6 002   A2 A3 B52 B3…               17 A*02:01g A*0…  0.0000000542         0.151</span></span>
<span><span class="co">#&gt; # ℹ 7 more variables: multiplier &lt;dbl&gt;, phased_freq &lt;dbl&gt;, phased_prob &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   haplo_freq_1 &lt;dbl&gt;, haplo_freq_2 &lt;dbl&gt;, haplo_rank_1 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   haplo_rank_2 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="working-with-multiple-allele-codes">Working with multiple allele codes<a class="anchor" aria-label="anchor" href="#working-with-multiple-allele-codes"></a>
</h4>
<p>py-ard can once again be used to lookup and decode <a href="https://hml.nmdp.org/MacUI/" class="external-link">Multiple Allele Codes (MACs)</a></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/mac_lookup.html">mac_lookup</a></span><span class="op">(</span><span class="va">ard</span>, <span class="st">"B*08:01/B*08:19N/B*08:109"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "B*08:YETY"</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/mac_decode.html">mac_decode</a></span><span class="op">(</span><span class="va">ard</span>, <span class="st">"B*08:YETY"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "B*08:01/B*08:19N/B*08:109"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="converting-to-and-from-gl-strings">Converting to and from GL Strings<a class="anchor" aria-label="anchor" href="#converting-to-and-from-gl-strings"></a>
</h4>
<p>Typing data often comes in a data frame like this:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing_df</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"001"</span>, <span class="st">"002"</span><span class="op">)</span>,</span>
<span>  A_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*01:01"</span>, <span class="st">"A*02:01"</span><span class="op">)</span>, A_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*03:01"</span>, <span class="st">"A*29:02"</span><span class="op">)</span>,</span>
<span>  B_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B*08:01"</span>, <span class="st">"B*07:02"</span><span class="op">)</span>, B_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B*07:02"</span>, <span class="cn">NA_character_</span><span class="op">)</span>,</span>
<span>  C_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C*07:01"</span>, <span class="st">"C*05:01"</span><span class="op">)</span>, C_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C*07:02"</span>, <span class="cn">NA_character_</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">typing_df</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 7</span></span>
<span><span class="co">#&gt;   id    A_1     A_2     B_1     B_2     C_1     C_2    </span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;  </span></span>
<span><span class="co">#&gt; 1 001   A*01:01 A*03:01 B*08:01 B*07:02 C*07:01 C*07:02</span></span>
<span><span class="co">#&gt; 2 002   A*02:01 A*29:02 B*07:02 &lt;NA&gt;    C*05:01 &lt;NA&gt;</span></span></code></pre></div>
<p>Use <code><a href="reference/df_to_gl.html">df_to_gl()</a></code> to convert all the individual allele columns into a single <a href="https://glstring.org" class="external-link">GL String</a>:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing_df_gl</span> <span class="op">&lt;-</span> <span class="va">typing_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># convert each typing to its own GL String</span></span>
<span>  <span class="fu"><a href="reference/df_to_gl.html">df_to_gl</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">typing_df_gl</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 2</span></span>
<span><span class="co">#&gt;   id    glstring                                                                </span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;                                                                   </span></span>
<span><span class="co">#&gt; 1 001   hla#2025-02-20#HLA-A*01:01+HLA-A*03:01^HLA-B*07:02+HLA-B*08:01^HLA-C*07…</span></span>
<span><span class="co">#&gt; 2 002   hla#2025-02-20#HLA-A*02:01+HLA-A*29:02^HLA-B*07:02^HLA-C*05:01</span></span></code></pre></div>
<p>Use <code><a href="reference/gl_to_df.html">gl_to_df()</a></code> to go the opposite way: from a dataframe of GL Strings to one columns per allele:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing_df_gl</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>gl_df <span class="op">=</span> <span class="fu"><a href="reference/gl_to_df.html">gl_to_df</a></span><span class="op">(</span><span class="va">glstring</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/unnest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">gl_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 11</span></span>
<span><span class="co">#&gt;   id    glstring      glstring_index namespace version_or_date A_1   A_2   B_1  </span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;                  &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;           &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;</span></span>
<span><span class="co">#&gt; 1 001   hla#2025-02-…              1 hla       2025-02-20      HLA-… HLA-… HLA-…</span></span>
<span><span class="co">#&gt; 2 002   hla#2025-02-…              2 hla       2025-02-20      HLA-… HLA-… HLA-…</span></span>
<span><span class="co">#&gt; # ℹ 3 more variables: B_2 &lt;chr&gt;, C_1 &lt;chr&gt;, C_2 &lt;chr&gt;</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="looking-up-eplets-and-alleles">Looking up eplets and alleles<a class="anchor" aria-label="anchor" href="#looking-up-eplets-and-alleles"></a>
</h4>
<p>Grab database from the <a href="https://www.epregistry.com.br" class="external-link">HLA Eplet registry</a> and use it to lookup which eplets occur on an HLA allele, or vice versa.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_eplets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_eplet_registry.html">load_eplet_registry</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loaded Eplet Registry table</span></span>
<span><span class="co">#&gt; IPD-IMGT/HLA version 3.59 (2025-02-01),</span></span>
<span><span class="co">#&gt; downloaded from https://www.epregistry.com.br</span></span>
<span><span class="fu"><a href="reference/lookup_alleles.html">lookup_alleles</a></span><span class="op">(</span><span class="va">df_eplets</span>, <span class="st">"17S"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`17S`</span></span>
<span><span class="co">#&gt; [1] "A*30:01" "A*30:02" "A*01:02"</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/lookup_eplets.html">lookup_eplets</a></span><span class="op">(</span><span class="va">df_eplets</span>, <span class="st">"A*01:02"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`A*01:02`</span></span>
<span><span class="co">#&gt;  [1] "9S"         "17S"        "44KM"       "62QE"       "65RA"      </span></span>
<span><span class="co">#&gt;  [6] "65RNA"      "66N"        "66NH"       "66NM"       "71HS"      </span></span>
<span><span class="co">#&gt; [11] "76ANT"      "77N[ABC]"   "77NGT"      "79GT"       "80T"       </span></span>
<span><span class="co">#&gt; [16] "80TL"       "90D"        "95I"        "97I"        "99Y"       </span></span>
<span><span class="co">#&gt; [21] "109F"       "114R"       "116D"       "138MI"      "144K"      </span></span>
<span><span class="co">#&gt; [26] "144KR"      "149AH"      "151H"       "152A"       "152HA"     </span></span>
<span><span class="co">#&gt; [31] "156R"       "163R"       "163RG"      "166DG"      "193PI"     </span></span>
<span><span class="co">#&gt; [36] "275EL"      "62QE+56G"   "79GT+90D"   "138MI+79GT" "144KR+151H"</span></span></code></pre></div>
<p>A common use case would be to lookup which eplets occur on a set of (positive) Luminex beads:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_eplets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_eplet_registry.html">load_eplet_registry</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loaded Eplet Registry table</span></span>
<span><span class="co">#&gt; IPD-IMGT/HLA version 3.59 (2025-02-01),</span></span>
<span><span class="co">#&gt; downloaded from https://www.epregistry.com.br</span></span>
<span><span class="va">luminex_df</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">sampleID</span>, <span class="op">~</span><span class="va">allele</span>, <span class="op">~</span><span class="va">positive</span>,</span>
<span>  <span class="st">"001"</span>, <span class="st">"A*01:01"</span>, <span class="cn">TRUE</span>,</span>
<span>  <span class="st">"001"</span>, <span class="st">"A*02:01"</span>, <span class="cn">FALSE</span>,</span>
<span>  <span class="st">"002"</span>, <span class="st">"A*01:01"</span>, <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/get_positive_eplets.html">get_positive_eplets</a></span><span class="op">(</span><span class="va">luminex_df</span>, <span class="va">sampleID</span>, <span class="va">allele</span>, <span class="va">positive</span>, <span class="va">df_eplets</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 28 × 2</span></span>
<span><span class="co">#&gt;    sampleID eplets_pos</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt;     </span></span>
<span><span class="co">#&gt;  1 001      44KM      </span></span>
<span><span class="co">#&gt;  2 001      62QE      </span></span>
<span><span class="co">#&gt;  3 001      65RNA     </span></span>
<span><span class="co">#&gt;  4 001      66N       </span></span>
<span><span class="co">#&gt;  5 001      66NH      </span></span>
<span><span class="co">#&gt;  6 001      66NM      </span></span>
<span><span class="co">#&gt;  7 001      76ANT     </span></span>
<span><span class="co">#&gt;  8 001      77N[ABC]  </span></span>
<span><span class="co">#&gt;  9 001      77NGT     </span></span>
<span><span class="co">#&gt; 10 001      90D       </span></span>
<span><span class="co">#&gt; # ℹ 18 more rows</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="antibodies">Antibodies<a class="anchor" aria-label="anchor" href="#antibodies"></a>
</h3>
<div class="section level4">
<h4 id="parsing-raw-luminex-files">Parsing raw Luminex files<a class="anchor" aria-label="anchor" href="#parsing-raw-luminex-files"></a>
</h4>
<p>Parse <code>.csv</code> files from Luminex Single-Antigen Bead (SAB) assays, along with the lot-specific <code>.eds</code> file that came with the kit, to produce interpreted results identical to Immucor’s MATCH IT!® Antibody Analysis Software:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># path to some mock data (originally for testing purposes)</span></span>
<span><span class="va">lum_path</span> <span class="op">&lt;-</span> <span class="fu">testthat</span><span class="fu">::</span><span class="fu"><a href="https://testthat.r-lib.org/reference/test_path.html" class="external-link">test_path</a></span><span class="op">(</span><span class="st">"luminex"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># parse the files</span></span>
<span><span class="fu"><a href="reference/read_lum_csv.html">read_lum_csv</a></span><span class="op">(</span></span>
<span>  csv_filepath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">lum_path</span>, <span class="st">"LSA1-test.csv"</span><span class="op">)</span>,</span>
<span>  lots_path <span class="op">=</span> <span class="va">lum_path</span> <span class="co"># folder with .eds file</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Location, Sample, antigen_id)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Location, Sample, antigen_id)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, antigen_id)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(antigen_id)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(antigen_id)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Location, Sample)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(antigen_id)`</span></span>
<span><span class="co">#&gt; # A tibble: 12 × 18</span></span>
<span><span class="co">#&gt;    Sample Count antigen_id Cutoff  Median mfi_lra assignment bg_adjusted ad_mfi</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt; &lt;chr&gt;       &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;            &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 S001      67 103          3.75  8002.   127.   Positive        7822.  6713. </span></span>
<span><span class="co">#&gt;  2 S001     104 PC          NA    18206.    36.3  Negative       16706.  4551. </span></span>
<span><span class="co">#&gt;  3 S001      87 NC          NA      161      1    Negative          NA     NA  </span></span>
<span><span class="co">#&gt;  4 S001      73 104          4.01    63      1    Negative        -107     52.1</span></span>
<span><span class="co">#&gt;  5 S001      89 105          3.79   127      1    Negative         -53     97.4</span></span>
<span><span class="co">#&gt;  6 S001      60 106          3.49   501      1    Negative         256    634. </span></span>
<span><span class="co">#&gt;  7 S002     120 PC          NA    17573    158.   Negative       16073   4393. </span></span>
<span><span class="co">#&gt;  8 S002      74 104          4.01    92.5    1.34 Negative         -77.5   76.4</span></span>
<span><span class="co">#&gt;  9 S002      92 NC          NA      113      1    Negative          NA     NA  </span></span>
<span><span class="co">#&gt; 10 S002      82 103          3.75    69      1    Negative        -111     57.9</span></span>
<span><span class="co">#&gt; 11 S002     113 105          3.79   120      1    Negative         -60     92.0</span></span>
<span><span class="co">#&gt; 12 S002      78 106          3.49   112.     1    Negative        -134.   141. </span></span>
<span><span class="co">#&gt; # ℹ 9 more variables: ad_bg_adjusted &lt;dbl&gt;, A &lt;chr&gt;, B &lt;chr&gt;, C &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   Bw &lt;chr&gt;, A_serology &lt;chr&gt;, B_serology &lt;chr&gt;, C_serology &lt;chr&gt;, RAD &lt;dbl&gt;</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="other-packages">Other packages<a class="anchor" aria-label="anchor" href="#other-packages"></a>
</h2>
<p>There’s many other implementations with partly overlapping goals.</p>
<ul>
<li>R:
<ul>
<li>
<a href="https://github.com/gschofl/hlatools" class="external-link">hlatools</a> provides access to the <a href="https://www.ebi.ac.uk/ipd/imgt/hla/" class="external-link">IPD-IMGT/HLA database</a>
</li>
<li>
<a href="https://cran.r-project.org/web/packages/hlaR" class="external-link">hlaR</a> does cleaning and imputation of HLA typings</li>
<li>
<a href="http://bioconductor.org/packages/release/bioc/html/immunotation.html" class="external-link">immunotation</a> formats HLA typings and can work with haplotype/allele frequencies</li>
</ul>
</li>
<li>Python:
<ul>
<li>
<a href="https://github.com/nmdp-bioinformatics/py-ard" class="external-link">py-ard</a> reduces HLA typing resolution. hlapro actually depends on <code>py-ard</code> for some of this functionality.</li>
<li>
<a href="https://github.com/nmdp-bioinformatics/pyglstring" class="external-link">pyglstring</a> checks whether GL Strings are well formatted</li>
<li>
<a href="https://github.com/lgragert/hla-who-to-unos" class="external-link">ALLAN</a> converts between serological and molecular HLA notation</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="why-hlapro">Why <code>hlapro</code>?<a class="anchor" aria-label="anchor" href="#why-hlapro"></a>
</h2>
<p>This package is developed mainly for use in the <a href="https://doi.org/10.1111/tan.13581" class="external-link"><strong>PRO</strong>CARE</a> project, which (among others) aims to <strong>pro</strong>file (HLA-)antibodies and their role in kidney transplant rejection. hlapro aims to be a one-stop shop for all the analyses in the second iteration of PROCARE.</p>
</div>
<div class="section level2">
<h2 id="attribution">Attribution<a class="anchor" aria-label="anchor" href="#attribution"></a>
</h2>
<p>Package logo made with <a href="https://connect.thinkr.fr/hexmake/" class="external-link">hexmake</a>; Icon by <a href="https://www.freepik.com/icon/test-tube_9583096#fromView=search&amp;term=hla+gene&amp;page=2&amp;position=35" class="external-link">Freepik</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/lcreteig/hlapro/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/lcreteig/hlapro/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing hlapro</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Leon Reteig <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5814-0992" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>UMC Utrecht <br><small class="roles"> Copyright holder, funder </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/lcreteig/hlapro/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/lcreteig/hlapro/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://github.com/lcreteig/hlapro/actions/workflows/pkgdown.yaml" class="external-link"><img src="https://github.com/lcreteig/hlapro/actions/workflows/pkgdown.yaml/badge.svg" alt="pkgdown"></a></li>
<li><a href="https://CRAN.R-project.org/package=hlapro" class="external-link"><img src="https://www.r-pkg.org/badges/version/hlapro" alt="CRAN status"></a></li>
<li><a href="https://app.codecov.io/gh/lcreteig/hlapro" class="external-link"><img src="https://codecov.io/gh/lcreteig/hlapro/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Leon Reteig, UMC Utrecht.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
