<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Provides tools for Human Leukocyte Antigen (HLA) data:
    genotypes as well as (Luminex) single-antigen bead-based assays.
    Including: standardizing HLA typings to a common format (GL Strings),
    reducing and increasing HLA typing resolution, determining bead
    positivity and assessing donor-specific antibodies (DSAs).">
<title>Tools to Work with HLA Typings and Single-Antigen Bead Assays • hlapro</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Tools to Work with HLA Typings and Single-Antigen Bead Assays">
<meta property="og:description" content="Provides tools for Human Leukocyte Antigen (HLA) data:
    genotypes as well as (Luminex) single-antigen bead-based assays.
    Including: standardizing HLA typings to a common format (GL Strings),
    reducing and increasing HLA typing resolution, determining bead
    positivity and assessing donor-specific antibodies (DSAs).">
<meta property="og:image" content="https://lcreteig.github.io/hlapro/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">hlapro</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/lcreteig/hlapro/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="hlapro-">hlapro 
<a class="anchor" aria-label="anchor" href="#hlapro-"></a>
</h1>
</div>
<!-- badges: start -->

<p>The goal of hlapro is to provide some tooling to work with <a href="https://en.wikipedia.org/wiki/Human_leukocyte_antigen" class="external-link">Human Leukocyte Antigen (HLA)</a> data, in the context of transplantation research:</p>
<ul>
<li>HLA typings</li>
<li>single bead assays (e.g. Luminex) for detecting HLA-antibodies</li>
</ul>
<p>The package is currently still in (heavy) development, but ultimately the goal is to support the following functionality:</p>
<ul>
<li>
<strong>HLA typings</strong>:
<ul>
<li>Obtaining HLA mismatches between transplant donor and recipient</li>
<li>Formatting into <a href="https://glstring.org" class="external-link">GL Strings</a>
</li>
<li>Downscaling (reducing) typing resolution</li>
<li>Upscaling (imputing) typing resolution, based on haplotype frequency tables</li>
</ul>
</li>
<li>
<strong>HLA antibody assays</strong>:
<ul>
<li>Determining which beads in the assay are positive, based on automated implementations of different methods</li>
<li>Assessing the presence of donor-specific antibodies (DSA)</li>
</ul>
</li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of hlapro from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"lcreteig/hlapro"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lcreteig/hlapro" class="external-link">hlapro</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="mismatches">Mismatches<a class="anchor" aria-label="anchor" href="#mismatches"></a>
</h3>
<p>Get mismatched HLAs from donor and recipient typings</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">donor_typing</span> <span class="op">&lt;-</span> <span class="st">"A1 A2 B5"</span></span>
<span><span class="va">recipient_typing</span> <span class="op">&lt;-</span> <span class="st">"A1 A3 B5 B12"</span></span>
<span></span>
<span><span class="fu"><a href="reference/get_mismatches.html">get_mismatches</a></span><span class="op">(</span><span class="va">donor_typing</span>, <span class="va">recipient_typing</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A2"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extracting-alleles">Extracting alleles<a class="anchor" aria-label="anchor" href="#extracting-alleles"></a>
</h3>
<p>Extract HLA alleles from a typing string</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing</span> <span class="op">&lt;-</span> <span class="st">"A1 A2 B7 B8 Cw3 DQ5 DQ8 DR4 DR11 DR52 DR53"</span></span>
<span><span class="fu"><a href="reference/extract_alleles_str.html">extract_alleles_str</a></span><span class="op">(</span><span class="va">typing</span><span class="op">)</span></span>
<span><span class="co">#&gt;    A_1    A_2    B_1    B_2    C_1    C_2 DPA1_1 DPA1_2 DPB1_1 DPB1_2 DQA1_1 </span></span>
<span><span class="co">#&gt;    "1"    "2"    "7"    "8"    "3"     NA     NA     NA     NA     NA     NA </span></span>
<span><span class="co">#&gt; DQA1_2 DQB1_1 DQB1_2 DRB1_1 DRB1_2 DRB._1 DRB._2 </span></span>
<span><span class="co">#&gt;     NA    "5"    "8"    "4"   "11"   "52"   "53"</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>typing <span class="op">=</span> <span class="st">"A1 A2 B7 B8 Cw3 DQ5 DQ8 DR4 DR11 DR52 DR53"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/extract_alleles_str.html">extract_alleles_df</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">typing</span>, loci <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(typing)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(typing)`</span></span>
<span><span class="co">#&gt; # A tibble: 1 × 7</span></span>
<span><span class="co">#&gt;   typing                                     A_1   A_2   B_1   B_2   C_1   C_2  </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                                      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;</span></span>
<span><span class="co">#&gt; 1 A1 A2 B7 B8 Cw3 DQ5 DQ8 DR4 DR11 DR52 DR53 1     2     7     8     3     ""</span></span></code></pre></div>
<p>If there’s more than two alleles for a given locus in the typing, you’ll receive a warning:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing</span> <span class="op">&lt;-</span> <span class="st">"A1 A2 A3 B7 B8 Cw1"</span></span>
<span><span class="fu"><a href="reference/extract_alleles_str.html">extract_alleles_str</a></span><span class="op">(</span><span class="va">typing</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: One or more loci found with more than 2 alleles.</span></span>
<span><span class="co">#&gt; ✖ `extract_alleles_str()` will only pick the first two.</span></span>
<span><span class="co">#&gt; ℹ Use `hlapro::count_alleles()` to find out more.</span></span>
<span><span class="co">#&gt;    A_1    A_2    B_1    B_2    C_1    C_2 DPA1_1 DPA1_2 DPB1_1 DPB1_2 DQA1_1 </span></span>
<span><span class="co">#&gt;    "1"    "2"    "7"    "8"    "1"     NA     NA     NA     NA     NA     NA </span></span>
<span><span class="co">#&gt; DQA1_2 DQB1_1 DQB1_2 DRB1_1 DRB1_2 DRB._1 DRB._2 </span></span>
<span><span class="co">#&gt;     NA     NA     NA     NA     NA     NA     NA</span></span></code></pre></div>
<p>Use <code><a href="reference/count_alleles.html">count_alleles()</a></code> to easily inspect the number of alleles per locus:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/count_alleles.html">count_alleles</a></span><span class="op">(</span><span class="va">typing</span><span class="op">)</span></span>
<span><span class="co">#&gt;    A    B    C DPA1 DPB1 DQA1 DQB1 DRB1 DRB. </span></span>
<span><span class="co">#&gt;    3    2    1    0    0    0    0    0    0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cleaning-allele-strings">Cleaning allele strings<a class="anchor" aria-label="anchor" href="#cleaning-allele-strings"></a>
</h3>
<p>Correcting some common formatting issues</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allele_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">" A'*1 "</span>, <span class="co"># spurious whitespace, punctuation, no leading zero, missing XX code</span></span>
<span>  <span class="st">"A10(25)"</span>, <span class="co"># redundant typing (includes both split and broad)</span></span>
<span>  <span class="st">"C*03:01/02"</span> <span class="co"># locus and allele group left off in ambiguity</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/clean_hla.html">clean_hla</a></span><span class="op">(</span><span class="va">allele_vec</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A*01:XX"         "A25"             "C*03:01/C*03:02"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="validating-alleles">Validating alleles<a class="anchor" aria-label="anchor" href="#validating-alleles"></a>
</h3>
<p>Check whether alleles are well-formed</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/validate_allele.html">validate_allele</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A2"</span>, <span class="st">"A*01:AABJE"</span>, <span class="st">"A*24:02:01:02L"</span>, <span class="st">"not-an-HLA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  TRUE  TRUE  TRUE FALSE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="allele-resolution">Allele resolution<a class="anchor" aria-label="anchor" href="#allele-resolution"></a>
</h3>
<p>Determine whether an allele is of low/intermediate/high resolution</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/get_resolution.html">get_resolution</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A2"</span>, <span class="st">"A*01:AABJE"</span>, <span class="st">"B*42:08"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "low"          "intermediate" "high"</span></span></code></pre></div>
<p>Use the extended mode to get more information:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/get_resolution.html">get_resolution</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A2"</span>, <span class="st">"A*24:XX"</span>, <span class="st">"A*01:AB"</span>, <span class="st">"B*42:08"</span>, <span class="st">"A*01:01:01:01"</span><span class="op">)</span>,</span>
<span>  extended <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "serology - broad"    "molecular - split"   "intermediate"       </span></span>
<span><span class="co">#&gt; [4] "high - second field" "high - fourth field"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="downscaling-to-serological-equivalents">Downscaling to serological equivalents<a class="anchor" aria-label="anchor" href="#downscaling-to-serological-equivalents"></a>
</h3>
<p>Get the serological equivalents of an allele as defined by the <a href="https://etrl.eurotransplant.org/resources/hla-tables/" class="external-link">ETRL HLA</a> conversion tables</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/get_serology.html">get_serology</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B*15:79N"</span>, <span class="st">"B*15:YETY"</span>, <span class="st">"B*15:01:16"</span>, <span class="st">"B*15:02"</span>, <span class="st">"B*15:85"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] NA    "B15" "B62" "B75" "B15"</span></span></code></pre></div>
<p>Also supports lookup of broads or splits specifically:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alleles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*01:01"</span>, <span class="st">"A*25:76:02"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/get_split.html">get_split</a></span><span class="op">(</span><span class="va">alleles</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] NA    "A25"</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A24"</span>, <span class="st">"A*25:76:02"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/get_broad.html">get_broad</a></span><span class="op">(</span><span class="va">splits</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A9"  "A10"</span></span></code></pre></div>
<p>And whether an allele has the Bw4 or Bw6 epitope:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B14"</span>, <span class="st">"B63"</span>, <span class="st">"B*40:05"</span>, <span class="st">"A*01:01"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/get_public.html">get_public</a></span><span class="op">(</span><span class="va">b_s</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Bw6" "Bw4" "Bw6" NA</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="upscaling-from-serological-equivalents-to-2-field-high-resolution">Upscaling from serological equivalents to 2-field high resolution<a class="anchor" aria-label="anchor" href="#upscaling-from-serological-equivalents-to-2-field-high-resolution"></a>
</h3>
<p>Depends on the haplotype frequencies released by the NMDP, which must be downloaded from <a href="https://frequency.nmdp.org" class="external-link">here</a> after logging in and accepting the license.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/upscale_typings.html">upscale_typings</a></span><span class="op">(</span></span>
<span>  filepath <span class="op">=</span> <span class="st">"~/Downloads/A~C~B~DRB3-4-5~DRB1~DQB1.xlsx"</span>,</span>
<span>  typing <span class="op">=</span> <span class="st">"A24 A28 B35 B61 DR4 DR11"</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">unphased_geno</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"haplo"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 5</span></span>
<span><span class="co">#&gt;   unphased_geno              haplo_freq_1 haplo_freq_2 haplo_rank_1 haplo_rank_2</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                             &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 A*24:02g A*68:01g B*35:03…    0.000585     0.000209           240          665</span></span>
<span><span class="co">#&gt; 2 A*24:02g A*68:01g B*35:03…    0.0000849    0.0000597         1565         2058</span></span></code></pre></div>
<p>Also able to upscale multiple typings at once, for instance in a dataframe:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing_df</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"001"</span>, <span class="st">"002"</span><span class="op">)</span>,</span>
<span>  input_typings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"A24 A28 B35 B61 DR4 DR11"</span>,</span>
<span>    <span class="st">"A2 A3 B52 B35 Cw4 DR11 DR52 DQ3"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">typing_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>geno_df <span class="op">=</span> <span class="fu"><a href="reference/upscale_typings.html">upscale_typings</a></span><span class="op">(</span></span>
<span>    <span class="st">"~/Downloads/A~C~B~DRB3-4-5~DRB1~DQB1.xlsx"</span>,</span>
<span>    <span class="va">input_typings</span>,</span>
<span>    as_list <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/unnest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">geno_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 13</span></span>
<span><span class="co">#&gt;   id    input_typings id_unphased_geno unphased_geno unphased_freq unphased_prob</span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;                    &lt;int&gt; &lt;chr&gt;                 &lt;dbl&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 001   A24 A28 B35 …              364 A*24:02g A*6…  0.000000254          0.121</span></span>
<span><span class="co">#&gt; 2 001   A24 A28 B35 …              364 A*24:02g A*6…  0.000000254          0.121</span></span>
<span><span class="co">#&gt; 3 002   A2 A3 B52 B3…               17 A*02:01g A*0…  0.0000000542         0.151</span></span>
<span><span class="co">#&gt; 4 002   A2 A3 B52 B3…               17 A*02:01g A*0…  0.0000000542         0.151</span></span>
<span><span class="co">#&gt; 5 002   A2 A3 B52 B3…               17 A*02:01g A*0…  0.0000000542         0.151</span></span>
<span><span class="co">#&gt; 6 002   A2 A3 B52 B3…               17 A*02:01g A*0…  0.0000000542         0.151</span></span>
<span><span class="co">#&gt; # ℹ 7 more variables: multiplier &lt;dbl&gt;, phased_freq &lt;dbl&gt;, phased_prob &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   haplo_freq_1 &lt;dbl&gt;, haplo_freq_2 &lt;dbl&gt;, haplo_rank_1 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   haplo_rank_2 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="converting-to-and-from-gl-strings">Converting to and from GL Strings<a class="anchor" aria-label="anchor" href="#converting-to-and-from-gl-strings"></a>
</h3>
<p>Typing data often comes in a data frame like this:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing_df</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"001"</span>, <span class="st">"002"</span><span class="op">)</span>,</span>
<span>  A_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*01:01"</span>, <span class="st">"A*02:01"</span><span class="op">)</span>, A_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*03:01"</span>, <span class="st">"A*29:02"</span><span class="op">)</span>,</span>
<span>  B_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B*08:01"</span>, <span class="st">"B*07:02"</span><span class="op">)</span>, B_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B*07:02"</span>, <span class="cn">NA_character_</span><span class="op">)</span>,</span>
<span>  C_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C*07:01"</span>, <span class="st">"C*05:01"</span><span class="op">)</span>, C_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C*07:02"</span>, <span class="cn">NA_character_</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">typing_df</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 7</span></span>
<span><span class="co">#&gt;   id    A_1     A_2     B_1     B_2     C_1     C_2    </span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;  </span></span>
<span><span class="co">#&gt; 1 001   A*01:01 A*03:01 B*08:01 B*07:02 C*07:01 C*07:02</span></span>
<span><span class="co">#&gt; 2 002   A*02:01 A*29:02 B*07:02 &lt;NA&gt;    C*05:01 &lt;NA&gt;</span></span></code></pre></div>
<p>Use <code><a href="reference/df_to_gl.html">df_to_gl()</a></code> to convert all the individual allele columns into a single <a href="https://glstring.org" class="external-link">GL String</a>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing_df_gl</span> <span class="op">&lt;-</span> <span class="va">typing_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># convert each typing to its own GL String</span></span>
<span>  <span class="fu"><a href="reference/df_to_gl.html">df_to_gl</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">typing_df_gl</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 2</span></span>
<span><span class="co">#&gt;   id    glstring                                                                </span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;                                                                   </span></span>
<span><span class="co">#&gt; 1 001   hla#2023-09-20#HLA-A*01:01+HLA-A*03:01^HLA-B*07:02+HLA-B*08:01^HLA-C*07…</span></span>
<span><span class="co">#&gt; 2 002   hla#2023-09-20#HLA-A*02:01+HLA-A*29:02^HLA-B*07:02^HLA-C*05:01</span></span></code></pre></div>
<p>Use <code><a href="reference/gl_to_df.html">gl_to_df()</a></code> to go the opposite way: from a dataframe of GL Strings to one columns per allele:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typing_df_gl</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>gl_df <span class="op">=</span> <span class="fu"><a href="reference/gl_to_df.html">gl_to_df</a></span><span class="op">(</span><span class="va">glstring</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/unnest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">gl_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 11</span></span>
<span><span class="co">#&gt;   id    glstring      glstring_index namespace version_or_date A_1   A_2   B_1  </span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;                  &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;           &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;</span></span>
<span><span class="co">#&gt; 1 001   hla#2023-09-…              1 hla       2023-09-20      HLA-… HLA-… HLA-…</span></span>
<span><span class="co">#&gt; 2 002   hla#2023-09-…              2 hla       2023-09-20      HLA-… HLA-… HLA-…</span></span>
<span><span class="co">#&gt; # ℹ 3 more variables: B_2 &lt;chr&gt;, C_1 &lt;chr&gt;, C_2 &lt;chr&gt;</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="looking-up-eplets-and-alleles">Looking up eplets and alleles<a class="anchor" aria-label="anchor" href="#looking-up-eplets-and-alleles"></a>
</h2>
<p>Grab database from the <a href="https://www.epregistry.com.br" class="external-link">HLA Eplet registry</a> and use it to lookup which eplets occur on an HLA allele, or vice versa.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_eplets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_eplet_registry.html">load_eplet_registry</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loaded Eplet Registry table (Updated with IPD-IMGT/HLA 3.53.),</span></span>
<span><span class="co">#&gt; released 2023-08-25, downloaded from https://www.epregistry.com.br</span></span>
<span><span class="fu"><a href="reference/lookup_alleles.html">lookup_alleles</a></span><span class="op">(</span><span class="va">df_eplets</span>, <span class="st">"9F"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`9F`</span></span>
<span><span class="co">#&gt;  [1] "A*01:01"    "A*02:01"    "A*02:02"    "A*02:03"    "A*02:07"   </span></span>
<span><span class="co">#&gt;  [6] "A*02:18"    "A*03:01"    "A*03:02"    "A*32:01"    "A*36:01"   </span></span>
<span><span class="co">#&gt; [11] "A*74:01"    "A*80:01"    "C*01:02"    "C*01:03"    "DQB1*04:01"</span></span>
<span><span class="co">#&gt; [16] "DQB1*04:02" "DQB1*06:02"</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/lookup_eplets.html">lookup_eplets</a></span><span class="op">(</span><span class="va">df_eplets</span>, <span class="st">"A*01:01"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`A*01:01`</span></span>
<span><span class="co">#&gt;  [1] "9F"    "44KM"  "62QE"  "65RA"  "65RNA" "66N"   "66NH"  "66NM"  "71HS" </span></span>
<span><span class="co">#&gt; [10] "76ANT" "77N"   "77NGT" "79GT"  "80T"   "80TL"  "90D"   "95I"   "97I"  </span></span>
<span><span class="co">#&gt; [19] "99Y"   "109F"  "114R"  "116D"  "138MI" "144K"  "144KR" "149AH" "151H" </span></span>
<span><span class="co">#&gt; [28] "152A"  "152HA" "156R"  "163R"  "163RG" "166DG" "193PI" "275EL"</span></span></code></pre></div>
<p>A common use case would be to lookup which eplets occur on a set of (positive) Luminex beads:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_eplets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_eplet_registry.html">load_eplet_registry</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loaded Eplet Registry table (Updated with IPD-IMGT/HLA 3.53.),</span></span>
<span><span class="co">#&gt; released 2023-08-25, downloaded from https://www.epregistry.com.br</span></span>
<span><span class="va">luminex_df</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">sampleID</span>, <span class="op">~</span><span class="va">allele</span>, <span class="op">~</span><span class="va">positive</span>,</span>
<span>  <span class="st">"001"</span>, <span class="st">"A*01:01"</span>, <span class="cn">TRUE</span>,</span>
<span>  <span class="st">"001"</span>, <span class="st">"A*02:01"</span>, <span class="cn">FALSE</span>,</span>
<span>  <span class="st">"002"</span>, <span class="st">"A*01:01"</span>, <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/get_positive_eplets.html">get_positive_eplets</a></span><span class="op">(</span><span class="va">luminex_df</span>, <span class="va">sampleID</span>, <span class="va">allele</span>, <span class="va">positive</span>, <span class="va">df_eplets</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 24 × 2</span></span>
<span><span class="co">#&gt;    sampleID eplets_pos</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt;     </span></span>
<span><span class="co">#&gt;  1 001      44KM      </span></span>
<span><span class="co">#&gt;  2 001      62QE      </span></span>
<span><span class="co">#&gt;  3 001      65RNA     </span></span>
<span><span class="co">#&gt;  4 001      66N       </span></span>
<span><span class="co">#&gt;  5 001      66NH      </span></span>
<span><span class="co">#&gt;  6 001      66NM      </span></span>
<span><span class="co">#&gt;  7 001      76ANT     </span></span>
<span><span class="co">#&gt;  8 001      77N       </span></span>
<span><span class="co">#&gt;  9 001      77NGT     </span></span>
<span><span class="co">#&gt; 10 001      90D       </span></span>
<span><span class="co">#&gt; # ℹ 14 more rows</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="other-packages">Other packages<a class="anchor" aria-label="anchor" href="#other-packages"></a>
</h2>
<p>There’s many other implementations with partly overlapping goals (some of which hlapro might depend on in the future).</p>
<ul>
<li>R:
<ul>
<li>
<a href="https://github.com/gschofl/hlatools" class="external-link">hlatools</a> provides access to the <a href="https://www.ebi.ac.uk/ipd/imgt/hla/" class="external-link">IPD-IMGT/HLA database</a>
</li>
<li>
<a href="https://cran.r-project.org/web/packages/hlaR" class="external-link">hlaR</a> does cleaning and imputation of HLA typings</li>
<li>
<a href="http://bioconductor.org/packages/release/bioc/html/immunotation.html" class="external-link">immunotation</a> formats HLA typings and can work with haplotype/allele frequencies</li>
</ul>
</li>
<li>Python:
<ul>
<li>
<a href="https://github.com/nmdp-bioinformatics/py-ard" class="external-link">py-ard</a> reduces HLA typing resolution</li>
<li>
<a href="https://github.com/nmdp-bioinformatics/pyglstring" class="external-link">pyglstring</a> checks whether GL Strings are well formatted</li>
<li>
<a href="https://github.com/lgragert/hla-who-to-unos" class="external-link">ALLAN</a> converts between serological and molecular HLA notation</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="why-hlapro">Why <code>hlapro</code>?<a class="anchor" aria-label="anchor" href="#why-hlapro"></a>
</h2>
<p>This package is developed mainly for use in the <a href="https://doi.org/10.1111/tan.13581" class="external-link"><strong>PRO</strong>CARE</a> project, which (among others) aims to <strong>pro</strong>file (HLA-)antibodies and their role in kidney transplant rejection. hlapro aims to be a one-stop shop for all the analyses in the second iteration of PROCARE.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/lcreteig/hlapro/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/lcreteig/hlapro/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing hlapro</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Leon Reteig <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5814-0992" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>UMC Utrecht <br><small class="roles"> Copyright holder, funder </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/lcreteig/hlapro/actions/workflows/check-standard.yaml" class="external-link"><img src="https://github.com/lcreteig/hlapro/actions/workflows/check-standard.yaml/badge.svg" alt="check-standard"></a></li>
<li><a href="https://github.com/lcreteig/hlapro/actions/workflows/pkgdown.yaml" class="external-link"><img src="https://github.com/lcreteig/hlapro/actions/workflows/pkgdown.yaml/badge.svg" alt="pkgdown"></a></li>
<li><a href="https://app.codecov.io/gh/lcreteig/hlapro?branch=main" class="external-link"><img src="https://codecov.io/gh/lcreteig/hlapro/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://CRAN.R-project.org/package=hlapro" class="external-link"><img src="https://www.r-pkg.org/badges/version/hlapro" alt="CRAN status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Leon Reteig, UMC Utrecht.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
