---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# hlapro <a href="https://lcreteig.github.io/hlapro/"><img src="man/figures/logo.png" align="right" height="139" alt="hlapro website" /></a>

<!-- badges: start -->

[![R-CMD-check](https://github.com/lcreteig/hlapro/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/lcreteig/hlapro/actions/workflows/R-CMD-check.yaml)
[![pkgdown](https://github.com/lcreteig/hlapro/actions/workflows/pkgdown.yaml/badge.svg)](https://github.com/lcreteig/hlapro/actions/workflows/pkgdown.yaml)
[![CRAN status](https://www.r-pkg.org/badges/version/hlapro)](https://CRAN.R-project.org/package=hlapro)
[![Codecov test coverage](https://codecov.io/gh/lcreteig/hlapro/graph/badge.svg)](https://app.codecov.io/gh/lcreteig/hlapro)
<!-- badges: end -->

The goal of hlapro is to provide some tooling to work with [Human Leukocyte Antigen (HLA)](https://en.wikipedia.org/wiki/Human_leukocyte_antigen) data, in the context of transplantation research:

-   HLA typings
-   single bead assays (e.g. Luminex) for detecting HLA-antibodies

The package is currently still in (heavy) development, but ultimately the goal is to support the following functionality:

-   **HLA typings**:
    -   Obtaining HLA mismatches between transplant donor and recipient
    -   Formatting into [GL Strings](https://glstring.org)
    -   Downscaling (reducing) typing resolution
    -   Upscaling (imputing) typing resolution, based on haplotype frequency tables
-   **HLA antibody assays**:
    -   Determining which beads in the assay are positive, based on automated implementations of different methods
    -   Assessing the presence of donor-specific antibodies (DSA)

## Installation

You can install the development version of hlapro from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("lcreteig/hlapro")
```

## Usage

```{r}
library(hlapro)
```

### Typings

#### Mismatches

Get mismatched HLAs from donor and recipient typings

```{r mismatch}
donor_typing <- "A1 A2 B5"
recipient_typing <- "A1 A3 B5 B12"

get_mismatches(donor_typing, recipient_typing)
```

#### Extracting alleles

Extract HLA alleles from a typing string

```{r extract_alleles}
typing <- "A1 A2 B7 B8 Cw3 DQ5 DQ8 DR4 DR11 DR52 DR53"
extract_alleles_str(typing)

df <- tidyr::tibble(typing = "A1 A2 B7 B8 Cw3 DQ5 DQ8 DR4 DR11 DR52 DR53")
extract_alleles_df(df, typing, loci = c("A", "B", "C"))
```

If there's more than two alleles for a given locus in the typing, you'll receive a warning:

```{r more than two alleles}
typing <- "A1 A2 A3 B7 B8 Cw1"
extract_alleles_str(typing)
```
Use `count_alleles()` to easily inspect the number of alleles per locus:

```{r count alleles}
count_alleles(typing)
```

#### Cleaning allele strings

Correcting some common formatting issues

```{r clean alleles}
allele_vec <- c(
  " A'*1 ", # spurious whitespace, punctuation, no leading zero, missing XX code
  "A10(25)", # redundant typing (includes both split and broad)
  "C*03:01/02" # locus and allele group left off in ambiguity
)

clean_hla(allele_vec)
```

`clean_hla()` can also convert typings in the pre-2010 (v2) nomenclature to the updated (v3) nomenclature:...

```{r convert v2 to v3}
clean_hla(c("Cw*030205", "A*2416", "DPB1*02BYVD", "B*35UMU"))
```

...and convert deleted/changed alleles to their new assignments:

```{r convert deleted}
clean_hla(c("C*03:12", "C*12:139", "A*02:100"))
```

#### Validating alleles

Check whether alleles are well-formed

```{r validate_alleles}
validate_allele(c("A2", "A*01:AABJE", "A*24:02:01:02L", "not-an-HLA"))
```

#### Allele resolution

Determine whether an allele is of low/intermediate/high resolution

```{r resolution}
get_resolution(c("A2", "A*01:AABJE", "B*42:08"))
```

Use the extended mode to get more information:

```{r resolution_extended}
get_resolution(c("A2", "A*24:XX", "A*01:AB", "B*42:08", "A*01:01:01:01"),
  extended = TRUE
)
```

#### Downscaling to serological equivalents

There's two ways to do this:

1. Using the [IPD-IMGT/HLA database](https://www.ebi.ac.uk/ipd/imgt/hla/) as implemented by the [NMDP](https://network.nmdp.org/services-support/bioinformatics-immunobiology/tools) in [py-ard](https://github.com/nmdp-bioinformatics/py-ard):

```{r downscaling py-ard}
ard <- db_initialize(path.expand("~/ipd_db")) # (download &) initialize database
reduce_to_serology(
  ard,
  c("B*15:79N", "B*15:25/B*15:61", "B*15:02", "B*15:01:16")
)
```

2. Get the serological equivalents of an allele as defined by the [ETRL HLA](https://etrl.eurotransplant.org/resources/hla-tables/) conversion tables. This is less extensive and precise, but is nonetheless relevant for the Eurotransplant region.

```{r downscaling}
get_serology(c("B*15:79N", "B*15:25/B*15:61", "B*15:02", "B*15:01:16"))
```

The latter also supports lookup of broads or splits specifically:

```{r split}
alleles <- c("A*01:01", "A*25:76:02")
get_split(alleles)
```

```{r broad}
splits <- c("A24", "A*25:76:02")
get_broad(splits)
```

And whether an allele has the Bw4 or Bw6 epitope:

```{r public}
b_s <- c("B14", "B63", "B*40:05", "A*01:01")
get_public(b_s)
```

#### "Reducing" to two-field resolution

For many applications it suffices to work with HLA alleles at the protein-level. `py-ard` can also leverage the IPD-IMGT/HLA database to accomplish this:

```{r reduce to two fields}
reduce_to_field2(ard, c("A*01:01:01:01"))
```

The above case is arguably trivial, but many others are not:

```{r more two-field reductions}
reduce_to_field2(ard, "A*01:04:01:01N") # shortnull
reduce_to_field2(ard, "B*15:AH") # MACs
reduce_to_field2(ard, "Cw10") # serology
```

#### Upscaling from serological equivalents to 2-field high resolution

Depends on the haplotype frequencies released by the NMDP, which must be downloaded from [here](https://frequency.nmdp.org) after logging in and accepting the license.

```{r upscale typing} 
upscale_typings(
  filepath = "~/Downloads/A~C~B~DRB3-4-5~DRB1~DQB1.xlsx",
  typing = "A24 A28 B35 B61 DR4 DR11"
) |>
  dplyr::select(unphased_geno, dplyr::starts_with("haplo"))
```

Also able to upscale multiple typings at once, for instance in a dataframe:

```{r upscale df}
typing_df <- tidyr::tibble(
  id = c("001", "002"),
  input_typings = c(
    "A24 A28 B35 B61 DR4 DR11",
    "A2 A3 B52 B35 Cw4 DR11 DR52 DQ3"
  )
)
typing_df |>
  dplyr::mutate(geno_df = upscale_typings(
    "~/Downloads/A~C~B~DRB3-4-5~DRB1~DQB1.xlsx",
    input_typings,
    as_list = TRUE
  )) |>
  tidyr::unnest(geno_df)
```

#### Working with multiple allele codes

py-ard can once again be used to lookup and decode [Multiple Allele Codes (MACs)](https://hml.nmdp.org/MacUI/)

```{r encode mac}
mac_lookup(ard, "B*08:01/B*08:19N/B*08:109")
```

```{r lookup mac}
mac_decode(ard, "B*08:YETY")
```

#### Converting to and from GL Strings

Typing data often comes in a data frame like this:

```{r example typing df}
typing_df <- tidyr::tibble(
  id = c("001", "002"),
  A_1 = c("A*01:01", "A*02:01"), A_2 = c("A*03:01", "A*29:02"),
  B_1 = c("B*08:01", "B*07:02"), B_2 = c("B*07:02", NA_character_),
  C_1 = c("C*07:01", "C*05:01"), C_2 = c("C*07:02", NA_character_)
)
typing_df
```

Use `df_to_gl()` to convert all the individual allele columns into a single [GL String](https://glstring.org):

```{r df_to_gl}
typing_df_gl <- typing_df |>
  dplyr::group_by(id) |> # convert each typing to its own GL String
  df_to_gl()
typing_df_gl
```

Use `gl_to_df()` to go the opposite way: from a dataframe of GL Strings to one columns per allele:

```{r gl_to_df}
typing_df_gl |>
  dplyr::mutate(gl_df = gl_to_df(glstring)) |>
  tidyr::unnest(gl_df)
```

#### Looking up eplets and alleles

Grab database from the [HLA Eplet registry](https://www.epregistry.com.br) and use it to lookup which eplets occur on an HLA allele, or vice versa.

```{r lookup_alleles}
df_eplets <- load_eplet_registry()
lookup_alleles(df_eplets, "17S")
```

```{r lookup_eplets}
lookup_eplets(df_eplets, "A*01:02")
```
A common use case would be to lookup which eplets occur on a set of (positive) Luminex beads:

```{r}
df_eplets <- load_eplet_registry()
luminex_df <- dplyr::tribble(
  ~sampleID, ~allele, ~positive,
  "001", "A*01:01", TRUE,
  "001", "A*02:01", FALSE,
  "002", "A*01:01", FALSE
)
get_positive_eplets(luminex_df, sampleID, allele, positive, df_eplets)
```

### Antibodies

#### Parsing raw Luminex files

Parse `.csv` files from Luminex Single-Antigen Bead (SAB) assays, along with the lot-specific `.eds` file that came with the kit, to produce interpreted results identical to Immucor's MATCH IT!Â® Antibody Analysis Software:

```{r read luminex file}
# path to some mock data (originally for testing purposes)
lum_path <- testthat::test_path("luminex")

# parse the files
read_lum_csv(
  csv_filepath = file.path(lum_path, "LSA1-test.csv"),
  lots_path = lum_path # folder with .eds file
)
```

## Other packages

There's many other implementations with partly overlapping goals.

- R:
  - [hlatools](https://github.com/gschofl/hlatools) provides access to the [IPD-IMGT/HLA database](https://www.ebi.ac.uk/ipd/imgt/hla/)
  - [hlaR](https://cran.r-project.org/web/packages/hlaR) does cleaning and imputation of HLA typings
  - [immunotation](http://bioconductor.org/packages/release/bioc/html/immunotation.html) formats HLA typings and can work with haplotype/allele frequencies

- Python:
  - [py-ard](https://github.com/nmdp-bioinformatics/py-ard) reduces HLA typing resolution. hlapro actually depends on `py-ard` for some of this functionality.
  - [pyglstring](https://github.com/nmdp-bioinformatics/pyglstring) checks whether GL Strings are well formatted
  - [ALLAN](https://github.com/lgragert/hla-who-to-unos) converts between serological and molecular HLA notation

## Why `hlapro`?

This package is developed mainly for use in the [**PRO**CARE](https://doi.org/10.1111/tan.13581) project, which (among others) aims to **pro**file (HLA-)antibodies and their role in kidney transplant rejection. hlapro aims to be a one-stop shop for all the analyses in the second iteration of PROCARE.

## Attribution
Package logo made with [hexmake](https://connect.thinkr.fr/hexmake/); Icon by [Freepik](https://www.freepik.com/icon/test-tube_9583096#fromView=search&term=hla+gene&page=2&position=35).
