<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='upscale_typings() takes in a (vector of) low-resolution HLA typing
string(s), and "upscales" it to(two-field) high resolution based on
haplotype frequencies from the NMDP registry.'><title>Impute the resolution of an HLA typing from low to high (second field) — upscale_typings • hlapro</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Impute the resolution of an HLA typing from low to high (second field) — upscale_typings"><meta property="og:description" content='upscale_typings() takes in a (vector of) low-resolution HLA typing
string(s), and "upscales" it to(two-field) high resolution based on
haplotype frequencies from the NMDP registry.'><meta property="og:image" content="https://lcreteig.github.io/hlapro/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">hlapro</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/lcreteig/hlapro/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Impute the resolution of an HLA typing from low to high (second field)</h1>
      <small class="dont-index">Source: <a href="https://github.com/lcreteig/hlapro/blob/HEAD/R/upscaling.R" class="external-link"><code>R/upscaling.R</code></a></small>
      <div class="d-none name"><code>upscale_typings.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>upscale_typings()</code> takes in a (vector of) low-resolution HLA typing
string(s), and "upscales" it to(two-field) high resolution based on
<a href="http://dx.doi.org/10.1016/j.humimm.2013.06.025" class="external-link">haplotype frequencies from the NMDP registry</a>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">upscale_typings</span><span class="op">(</span></span>
<span>  <span class="va">filepath</span>,</span>
<span>  <span class="va">typings</span>,</span>
<span>  loci_input <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"DRB1"</span>, <span class="st">"DRB."</span>, <span class="st">"DQB1"</span><span class="op">)</span>,</span>
<span>  loci_output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"DRB1"</span>, <span class="st">"DRB."</span>, <span class="st">"DQB1"</span><span class="op">)</span>,</span>
<span>  population <span class="op">=</span> <span class="st">"EURCAU"</span>,</span>
<span>  n_haplos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_genos <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  as_list <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>filepath</dt>
<dd><p>String with the path to the <code>HLA-A~C~B~DRB1~DQB1.xlsx</code> file
as downloaded from <a href="https://frequency.nmdp.org" class="external-link">https://frequency.nmdp.org</a>.</p></dd>


<dt>typings</dt>
<dd><p>String or character vector of HLA typings, with
space-separated alleles in serological notation.</p></dd>


<dt>loci_input</dt>
<dd><p>Character vector of loci in the input genotype to be used
for upscaling. Must be a subset of <code>c("A", "B", "C", "DRB1", "DRB.", "DQB1")</code>, where <code>DRB.</code> is DRB3/4/5. If one of these loci does not occur in
the input typing, it will be ignored.</p></dd>


<dt>loci_output</dt>
<dd><p>Character vector of loci that the upscaled, output
genotype should contain. Can be different from <code>loci_input</code>, for example if
there's no typing at all for a certain locus and you want to infer it from
the haplotypes. Must be a subset of <code>c("A", "B", "C", "DRB1", "DRB.", "DQB1")</code>, where <code>DRB.</code> is DRB3/4/5.</p></dd>


<dt>population</dt>
<dd><p>String specifying which population group to use the
haplotype frequencies from. Must correspond to one of the abbreviations
that can be found on <a href="https://haplostats.org" class="external-link">https://haplostats.org</a>
Defaults to <code>"EURCAU"</code> (the largest population in the dataset).</p></dd>


<dt>n_haplos</dt>
<dd><p>Number of most frequent haplotypes to use for the upscaling,
e.g. <code>5000</code> to consider only the 5000 haplotypes with the highest frequency
(the rest is discarded). Defaults to all haplotypes with a non-zero
frequency in the selected population.</p></dd>


<dt>n_genos</dt>
<dd><p>Number of output genotypes to return for each input genotype,
sorted by probability (frequency) of the output genotypes. Defaults to the
most likely genotype only (i.e., <code>1</code>).</p></dd>


<dt>as_list</dt>
<dd><p>Boolean (<code>TRUE</code> or <code>FALSE</code>) that determines whether to return
the result as a single dataframe, or a list of dataframes: one for each
input typing (the latter can be useful when the input typings also live in
a data frame; see examples).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A (list of) dataframe(s) with the upscaling results, containing the
following columns:</p><ol><li><p><code>id_input_typing</code>: Sequential identifier for the input typing</p></li>
<li><p><code>id_unphased_geno</code>: Unique identifier for each output unphased genotype</p></li>
<li><p><code>unphased_geno</code>: Upscaled genotype</p></li>
<li><p><code>unphased_freq</code>: Frequency of upscaled genotype</p></li>
<li><p><code>unphased_prob</code>: Probability of upscaled genotype</p></li>
<li><p><code>phased_freq</code>: Frequency of phased genotype (= pair of haplotypes) that
make up the unphased genotype (can be many-to-one)</p></li>
<li><p><code>phased_prob</code>: Probability of the phased genotype</p></li>
<li><p><code>haplo_freq_1</code>: Frequency of the first haplotype in the phased genotype</p></li>
<li><p><code>haplo_rank_1</code>: Rank (descending) of the frequency of this haplotype</p></li>
<li><p><code>haplo_freq_2</code>: Frequency of the 2nd haplotype in the phased genotype</p></li>
<li><p><code>haplo_rank_2</code>: Rank (descending) of the frequency of this haplotype</p></li>
</ol></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function uses haplotype frequencies published by the NMDP at
<a href="https://frequency.nmdp.org" class="external-link">https://frequency.nmdp.org</a>. You'll need to login and accept the license to
obtain the data, which is why we cannot distribute it with this package.</p><div class="section">
<h3 id="imputation-algorithm">Imputation algorithm<a class="anchor" aria-label="anchor" href="#imputation-algorithm"></a></h3>


<p>Roughly, the function performs the following steps:</p><ol><li><p>Translate all haplotype alleles from high-resolution to their serological
equivalents</p></li>
<li><p>Select compatible haplotypes, i.e. those haplotypes with alleles that
<em>all</em> occur in the input genotype</p></li>
<li><p>Combine all compatible haplotypes into phased genotypes (i.e. unique
haplotype pairs). Fully homozygous haplotypes are not considered</p></li>
<li><p>Calculate the frequency and probability of the phased genotypes</p></li>
<li><p>Combine phased genotypes into unique unphased genotypes, and calculate
their frequency and probability</p></li>
</ol><p>For a detailed explanation of the terminology and imputation algorithm,
see the following references:</p><ul><li><p>Geffard et. al., Easy-HLA: a validated web application suite to reveal the
full details of HLA typing, <em>Bioinformatics, Volume 36, Issue 7</em>, April
2020, Pages 2157-2164, <a href="https://doi.org/10.1093/bioinformatics/btz875" class="external-link">https://doi.org/10.1093/bioinformatics/btz875</a></p></li>
<li><p>Madbouly, A., Gragert, L., Freeman, J., Leahy, N., Gourraud, P.-.-A.,
Hollenbach, J.A., Kamoun, M., Fernandez-Vina, M. and Maiers, M. (2014),
Validation of statistical imputation of allele-level multilocus phased
genotypes from ambiguous HLA assignments. <em>Tissue Antigens, 84</em>: 285-292.
<a href="https://doi.org/10.1111/tan.12390" class="external-link">https://doi.org/10.1111/tan.12390</a></p></li>
</ul></div>

    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">upscale_typings</span><span class="op">(</span></span></span>
<span class="r-in"><span>  filepath <span class="op">=</span> <span class="st">"~/Downloads/A~C~B~DRB3-4-5~DRB1~DQB1.xlsx"</span>,</span></span>
<span class="r-in"><span>  typing <span class="op">=</span> <span class="st">"A24 A28 B35 B61 DR4 DR11"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># If your GL Strings are in a data frame with some ID'ing columns that you</span></span></span>
<span class="r-in"><span><span class="co"># If you've got more than one typing to upscale, perhaps along with some</span></span></span>
<span class="r-in"><span><span class="co"># ID'ing columns (e.g. patient ID), it's probably best to put them in a</span></span></span>
<span class="r-in"><span><span class="co"># data frame and call `upscale_typings()` on your data frame:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">typing_df</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span></span></span>
<span class="r-in"><span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"001"</span>, <span class="st">"002"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  input_typings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"A24 A28 B35 B61 DR4 DR11"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"A2 A3 B52 B35 Cw4 DR11 DR52 DQ3"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">typing_df</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu">mutate</span><span class="op">(</span>geno_df <span class="op">=</span> <span class="fu">upscale_typings</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"~/Downloads/A~C~B~DRB3-4-5~DRB1~DQB1.xlsx"</span>,</span></span>
<span class="r-in"><span>    <span class="va">input_typings</span>,</span></span>
<span class="r-in"><span>    as_list <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span>  <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu">unnest</span><span class="op">(</span><span class="va">geno_df</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Leon Reteig, UMC Utrecht.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

